lib/xander/transaction.ex

```diff
@@ -0,0 +1,295 @@
+  def connected(
+        :internal,
+        :establish,
+        %__MODULE__{client: client, socket: socket, network: network} = data
+      ) do
+    Logger.info("Establishing handshake...")
+
+    :ok =
+      client.send(
+        socket,
+        Handshake.Proposal.version_message(@active_n2c_versions, network)
+      )
+
+    case client.recv(socket, 0, _timeout = 5_000) do
+      {:ok, full_response} ->
+        case Handshake.Response.validate(full_response) do
+          {:ok, _handshake_response} ->
+            # Transitions to idle state
+            Logger.info("Handshake successful")
+            {:next_state, :idle, data}
+
+          {:refused, response} ->
+            Logger.error("Handshake refused by node. Response: #{inspect(response)}")
+
+            {:next_state, :disconnected, data}
+        end
+
+      {:error, reason} ->
+        Logger.error("Error establishing connection #{inspect(reason)}")
+        {:next_state, :disconnected, data}
+    end
+  end
+
+  @spec busy(:info | :internal, {:send_tx, binary()} | {:tcp_closed, any()}, t()) ::
+          {:next_state, :disconnected, t()}
+          | {:next_state, :disconnected | :idle, t(), [{:reply, any(), {:error | :ok, any()}}]}
+  def busy(_event_type, _event_content, _data)
+
+  def busy(
+        :internal,
+        {:send_tx, tx_hex},
+        %__MODULE__{client: client, socket: socket, caller: from} = data
+      ) do
+    :ok = client.send(socket, Messages.transaction(tx_hex))
+
+    {next_state, actions} =
+      case client.recv(socket, 0, _timeout = 5_000) do
+        {:ok, tx_response} ->
+          handle_tx_response(tx_response, from)
+
+        {:error, reason} ->
+          Logger.error("Failed to send transaction: #{inspect(reason)}")
+          # Reply to caller with error
+          actions = [{:reply, from, {:error, {:send_failed, reason}}}]
+
+          {:idle, actions}
+      end
+
+    data = Map.delete(data, :caller)
+    {:next_state, next_state, data, actions}
+  end
+
+  def busy(:info, {:tcp_closed, socket}, %__MODULE__{socket: socket} = data) do
+    Logger.error("Connection closed while in busy")
+    {:next_state, :disconnected, data}
+  end
+
+  @doc """
+  Emits events when in the `idle` state.
+  This maps to the StIdle state in the Cardano Local State Query protocol.
+  This is the state where the process waits for a query to be made.
+  """
+  @spec idle(
+          :info | :internal | {:call, any()},
+          {:request, atom(), atom()} | {:tcp_closed, any()},
+          t()
+        ) ::
+          {:next_state, :disconnected, t()}
+          | {:next_state, :busy, t(), [{:next_event, :internal, {:send_tx, binary()}}]}
+  def idle(_event_type, _event_content, _data)
+
+  def idle(
+        {:call, from},
+        {:request, :send_tx, tx_hash},
+        %__MODULE__{client: _client, socket: _socket} = data
+      ) do
+    # Store the caller in data and don't reply immediately
+    data = Map.put(data, :caller, from)
```

Comment:
The use of a map here is a potential race condition. If a second call to this function comes in before [the function on line 182](https://github.com/wowica/xander/pull/21/files#diff-9bbbcd154b50aa7f7a1946fe8f5534ce7db56199b1a5b6e65f40035743da0d30R185) starts executing, then the first caller will be overwritten by the second one.

We should use a `queue` similar to how we're doing in [Xander.Query](https://github.com/wowica/xander/blob/main/lib/xander/query.ex#L172)

Reply:
I've made these updates. Note that the state function signatures needed to change as well, so some scrutiny is probably warranted!